### ISSUE TITLE
Repo: Investigate better handling of unconstrained generics

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10438]
### Suggestion

Sadly, right now `getConstrainedTypeOfLocation` is bugged for unconstrained generics. The code just gives back the type parameter type if getBaseConstraintOfType(nodeType) is nullish, but that's not correct. Unconstrained generics should behave like `unknown`. Here's what it is right now:

https://github.com/typescript-eslint/typescript-eslint/blob/62d57559564fb08512eafe03a2c1b167c4377601/packages/type-utils/src/getConstrainedTypeAtLocation.ts#L7-L20

And here's what it "should" be: 

```ts
/**
 * Resolves the given node's type. Will resolve to the type's generic constraint, if it has one.
 */
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type {
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();

    return (
      checker.getBaseConstraintOfType(nodeType) ?? checker.getUnknownType()
    );
  }
  return nodeType;
}
```

Unfortunately, due to https://github.com/microsoft/TypeScript/issues/60475, there is no `checker.getUnknownType()` yet or for a while.

It kind of sucks, but I propose that we instead modify the implementation to something more like
```ts
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type | undefined { // <-- note the `| undefined`
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();
    return checker.getBaseConstraintOfType(nodeType) 
    /* in the future this can have ?? checker.getUnknownType() */;
  }
  return nodeType;
}
```
and force callers to handle the unconstrained case separately.

***

Note - I think we should also double-check (and add tests) for what happens when the constraint is `any`. I'm pretty sure `function f<T extends any>(x: T) {}` behaves as though `T` were `unknown` rather than `any`. So maybe we should really be doing

```ts
export function getConstrainedTypeAtLocation(
  services: ParserServicesWithTypeInformation,
  node: TSESTree.Node,
): ts.Type | undefined {
  const nodeType = services.getTypeAtLocation(node);
  if (tsutils.isTypeParameter(nodeType)) {
    const checker = services.program.getTypeChecker();
    const constraint = checker.getBaseConstraintOfType(nodeType);
    return constraint == null || isTypeAnyType(constraint)
      ? /* in the future this can be checker.getUnknownType() */ undefined
      : constraint;
  }
  return nodeType;
}

```

I haven't yet checked what `checker.getBaseConstraintOfType()` does with an `any`.

### Additional Info

Related, caused https://github.com/typescript-eslint/typescript-eslint/issues/10311
Related, discussion at https://github.com/typescript-eslint/typescript-eslint/pull/10314#discussion_r1836097699
Possibly related, https://github.com/typescript-eslint/typescript-eslint/issues/10415

cc @ronami

**Comment by ronami** [comment]
Linking #10442, #10443 too (still haven't been triaged though).
